contrib/mesos/pkg/executor/service/service.go

```diff
@@ -86,188 +76,164 @@ func (s *KubeletExecutorServer) AddFlags(fs *pflag.FlagSet) {
+		// apply Messo specific settings
+		executorDone := make(chan struct{})
+		kcfg.Builder = func(kc *app.KubeletConfig) (app.KubeletBootstrap, *kconfig.PodConfig, error) {
+			k, pc, err := app.CreateAndInitKubelet(kc)
+			if err != nil {
+				return k, pc, err
+			}
 
-	diskSpacePolicy := kubelet.DiskSpacePolicy{
-		DockerFreeDiskMB: s.LowDiskSpaceThresholdMB,
-		RootFreeDiskMB:   s.LowDiskSpaceThresholdMB,
-	}
+			klet := k.(*kubelet.Kubelet)
 
-	//TODO(jdef) intentionally NOT initializing a cloud provider here since:
-	//(a) the kubelet doesn't actually use it
-	//(b) we don't need to create N-kubelet connections to zookeeper for no good reason
-	//cloud := cloudprovider.InitCloudProvider(s.CloudProvider, s.CloudConfigFile)
-	//log.Infof("Successfully initialized cloud provider: %q from the config file: %q\n", s.CloudProvider, s.CloudConfigFile)
+			s.kletLock.Lock()
+			s.klet = klet
+			s.kletLock.Unlock()
 
-	hostNetworkSources, err := kubelet.GetValidatedSources(strings.Split(s.HostNetworkSources, ","))
-	if err != nil {
-		return err
-	}
+			// decorate kubelet such that it shuts down when the executor is
+			decorated := &executorKubelet{
+				Kubelet:      klet,
+				kubeletDone:  kubeletDone,
+				executorDone: executorDone,
+			}
 
-	hostPIDSources, err := kubelet.GetValidatedSources(strings.Split(s.HostPIDSources, ","))
-	if err != nil {
-		return err
-	}
+			return decorated, pc, nil
+		}
+		kcfg.DockerDaemonContainer = "" // don't move the docker daemon into a cgroup
+		kcfg.Hostname = kcfg.HostnameOverride
+		kcfg.KubeClient = apiclient
+		kcfg.NodeName = kcfg.HostnameOverride
+		kcfg.PodConfig = kconfig.NewPodConfig(kconfig.PodConfigNotificationIncremental, kcfg.Recorder) // override the default pod source
+		kcfg.StandaloneMode = false
+		kcfg.SystemContainer = "" // don't take control over other system processes.
+
+		// create main pod source
+		updates := kcfg.PodConfig.Channel(MESOS_CFG_SOURCE)
+		go func() {
+			// execUpdates will be closed by the executor on shutdown
+			defer close(executorDone)
+
+			for u := range execUpdates {
+				u.Source = MESOS_CFG_SOURCE
+				updates <- u
+			}
+		}()
+
+		// create static-pods directory file source
+		log.V(2).Infof("initializing static pods source factory, configured at path %q", staticPodsConfigPath)
+		fileSourceUpdates := kcfg.PodConfig.Channel(kubelet.FileSource)
+		kconfig.NewSourceFile(staticPodsConfigPath, kcfg.HostnameOverride, kcfg.FileCheckFrequency, fileSourceUpdates)
 
-	hostIPCSources, err := kubelet.GetValidatedSources(strings.Split(s.HostIPCSources, ","))
+		// run the kubelet, until execUpdates is closed
+		// NOTE: because kcfg != nil holds, the upstream Run function will not
+		//       initialize the cloud provider. We explicitly wouldn't want
+		//       that because then every kubelet instance would query the master
+		//       state.json which does not scale.
+		err = s.KubeletServer.Run(kcfg)
+	}
 	if err != nil {
-		return err
+		close(kubeletDone)
```

Comment:
TODO: our kubelet.Run() override also closes this chan, be careful when refactoring